70_FLIPFluids/app_resources/compute/prepareCellUpdate.comp.hlsl

```diff
@@ -0,0 +1,58 @@
+#include "../common.hlsl"
+#include "../gridUtils.hlsl"
+#include "../descriptor_bindings.hlsl"
+
+struct SPushConstants
+{
+    uint numElements;
+    uint pad;
+};
+
+[[vk::push_constant]] SPushConstants pc;
+
+[[vk::binding(b_pcuGridData, s_pcu)]]
+cbuffer GridData
+{
+    SGridData gridData;
+};
+
+[[vk::binding(b_pcuPInBuffer, s_pcu)]] RWStructuredBuffer<Particle> particleInBuffer;
+[[vk::binding(b_pcuPOutBuffer, s_pcu)]] RWStructuredBuffer<Particle> particleOutBuffer;
+
+[[vk::binding(b_pcuPairBuffer, s_pcu)]] RWStructuredBuffer<uint2> particleCellPairBuffer;
+[[vk::binding(b_pcuGridIDBuffer, s_pcu)]] RWStructuredBuffer<uint2> gridParticleIDBuffer;
+
+[numthreads(WorkgroupSize, 1, 1)]
+void makeParticleCellPairs(uint32_t3 ID : SV_DispatchThreadID)
+{
+    uint p_id = ID.x;
+    uint c_id = worldPosToFlatIdx(particleInBuffer[p_id].position.xyz, gridData);
+
+    particleCellPairBuffer[p_id] = uint2(c_id, p_id);
+}
```

Comment:
I presume this is to feed your radix sort with a key-value pair?

If we actually had time to polish this example, then I'd require "Kernel Fusion" which would mean you move the first dispatch of the radix sort into here as to avoid a pipeline barrier and a dispatch

Same for you `shuffleParticles` I guess